[DOCS] Add local dev setup instructions

leemthompo opened the pull request:
- Removes existing content in [Run Elasticsearch locally](https://www.elastic.co/guide/en/elasticsearch/reference/current/run-elasticsearch-locally.html) which does not account for CA certs, and is not possible to use via cURL or language clients for local testing. This content is in a no-man's land that doesn't make it easy to test Elasticsearch locally if you're not using Kibana, and is also not production-ready.
- Replaces that page with minimal local dev Docker instructions that are quickest way to get started, and will work for local notebook testing, devs who just want to use terminal etc.
- This setup has been used for a while in the [Search Labs tutorial](https://www.elastic.co/search-labs/tutorials/install-elasticsearch/docker) and should be more prominent in our main docs
- Moves these instructions into a new Quickstart folder, along with existing [Quickstart guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/getting-started.html).
  - Replaces the instructions in the self-managed tab of quickstart with link to new local dev instructions
  - Moving into separate folder removes confusion that arises when these non-production ready instructions lived alongside production installation pages. 

## URL preview

https://elasticsearch_bk_107913.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/quickstart.html

## Follow-up work

- [x] Update Elasticsearch repo README installation instructions

leemthompo committed:
[DOCS] Add local dev setup instructions

- Replace existing Run ES in Docker locally page, with simpler no-security local dev setup
- Move this file into Quickstart folder, along with existing quickstart guide
- Update self-managed instructions in Quickstart guide to use local dev approach

leemthompo committed:
Cleanup git history

miguelgrinberg commented on docs/reference/quickstart/run-elasticsearch-locally.asciidoc:
```diff
@@ -0,0 +1,141 @@
+[source,sh]
+----
+export ELASTIC_PASSWORD="<ES_PASSWORD>"  # password for "elastic" username
+export KIBANA_PASSWORD="<KIB_PASSWORD"   # Used _internally_ by Kibana, must be at last 6 characters long
```
You are missing the closing `>` in the Kibana variable.

leemthompo committed:
Remove comments, fix typo, simplify docker wording

leemthompo committed:
Comments to fix weird test issue

leemthompo committed:
TIL:No need for NOTCONSOLE if snippet doesn't contain curl

leemthompo committed:
Fix typo

joemcelroy commented on docs/reference/quickstart/run-elasticsearch-locally.asciidoc:
```diff
@@ -0,0 +1,122 @@
+
+[WARNING]
+====
+*DO NOT USE THESE INSTRUCTIONS FOR PRODUCTION DEPLOYMENTS*
```
i think its enough to have it at the top only. Im not sure really helps having it in two places.

leemthompo replied:
Done

joemcelroy commented on docs/reference/quickstart/run-elasticsearch-locally.asciidoc:
```diff
@@ -0,0 +1,122 @@
+====
+*DO NOT USE THESE INSTRUCTIONS FOR PRODUCTION DEPLOYMENTS*
+
+The instructions on this page are for local development only. Do not use these instructions for production deployments, because they are not secure.
```
How about we have another section at the bottom to show an example on how to connect? It would benefit just showing the url, username + password connect example so its super obvious whats the next step to use with the client.

leemthompo replied:
Done

leemthompo committed:
- **Minor Text Update**: Updated the reference to "Docker development environment" to "Docker dev environment".

- **Unreleased Version Warning**: Added warnings to indicate if the Elasticsearch version is unreleased, no Docker image is available.
- **Language Client Connection**: Added instructions for connecting to Elasticsearch using language clients, including:
  - Python example using `elasticsearch` client.
  - cURL example using basic authentication.
- **Production Deployment Guidance**: The previous warning about not using the setup for production was replaced with a discrete "Next Steps" section recommending using Elastic Cloud for production deployments.

- ** Add Tip Added**: A tip was added at the top of the file suggesting local development setup for testing

- **Tab Names Updated**: The "Self-managed" tab was renamed to "Local Dev (Docker)" for consistency with the new local development setup recommendations.

leemthompo committed:
Delete stray id

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

miguelgrinberg commented on docs/reference/tab-widgets/quick-start-install-widget.asciidoc:
```diff
@@ -5,14 +5,14 @@
             aria-controls="self-managed-tab-install"
             id="self-managed-install"
             tabindex="-1">
-      Self-managed
```
The api-call-widget.asciidoc also needs the "Self-managed" tab to be renamed "Local Dev (Docker)" for consistency.

leemthompo committed:
Uniform tab names

leemthompo committed:
Remove ca cert flag from curl call

leemthompo committed:
Update abbreviated title for clearer nav experience

leemthompo committed:
Make next steps clearer

leemthompo committed:
Delete illegitimate possessive determiner

miguelgrinberg reviewed (approved):
LGTM!

leemthompo commented on docs/reference/quickstart/index.asciidoc:
```diff
@@ -0,0 +1,10 @@
+
+Get started quickly with {es}.
+
+* Learn how to run {es} (and {kib}) for <<run-elasticsearch-locally,Local development>>.
```
```suggestion
* Learn how to run {es} (and {kib}) for <<run-elasticsearch-locally,local development>>.
```

leemthompo committed:
Fix illegitimate capitalization

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.14](https://github.com/elastic/elasticsearch/pull/108376)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/108376">](https://github.com/elastic/elasticsearch/pull/108376) |

serenachou commented:
Feedback : 

I like that we distill it down to 3 mandatory steps: set your environment variables, set up the docker network and run this docker command. However, I like how Qdrant has 2 command simplicity: to say: docker: pull to get the image and then the run - it seems super simple. I'd rather not have them jump over to to https://elasticsearch_bk_107913.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/current/install-elasticsearch.html because then it is another click to https://elasticsearch_bk_107913.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/8.13/docker.html 

You already have it in the header that you're showing them how to do this with Docker, just give them the command to grab the right docker image on the page 

<img width="1236" alt="image" src="https://github.com/elastic/elasticsearch/assets/13574083/7b6483c2-9b19-46a4-baa7-7869f4e2afda">

I like that we keep it simple for the local install - but then have a link to the quickstart guide. I think we need to add to the Quick Start guide the section on "adding data" a link to getting started on connectors and deploying the connectors.

leemthompo commented:
Noted. Let's check back next week and see if we can shave anything else here :)

maxjakob commented on docs/reference/quickstart/run-elasticsearch-locally.asciidoc:
```diff
@@ -0,0 +1,177 @@
+docker run -p 127.0.0.1:9200:9200 -d --name elasticsearch --network elastic-net \
+  -e ELASTIC_PASSWORD=$ELASTIC_PASSWORD \
+  -e "discovery.type=single-node" \
+  -e "xpack.security.http.ssl.enabled=false" \
```
🥰

maxjakob replied:
I love this change. Having SSL out of the way is a big deal. One question though: why didn't we go for full `xpack.security.enabled=false` for local testing setup without password?

miguelgrinberg replied:
This was a concesion we've made.

But I personally think this is a good set up. Every other database I've used (except sqlite!) uses this style of dev solution, where there is no TLS but you still have to authenticate. For someone developing an app that will eventually run in production I think it is good to have access to the auth system, since you will obviously need to have it during development to have your app ready for prod.

xeraa replied:
Not that this is a realistic change to happen but I feel like `discovery.type=single-node` should imply `xpack.security.http.ssl.enabled=false`. You want TLS in a distributed system but it's less important for a local installation (and why it's so much easier for PostgreSQL,... to just have a generated password and be done with it)

leemthompo replied:
> Not that this is a realistic change to happen but I feel like discovery.type=single-node should imply xpack.security.http.ssl.enabled=false.

true

joemcelroy commented:
Theres no reason why we cant have the commands on the homepage all in one and have a link to the quickstart guide. 

The commands could be highlighted as just es only, same way as qdrant (note here im missing kibana, this is just es only) 

```
export ELASTIC_PASSWORD="<ES_PASSWORD>" 
docker run -p 127.0.0.1:9200:9200 -d \
  -e ELASTIC_PASSWORD=$ELASTIC_PASSWORD \
  -e "discovery.type=single-node" \
  -e "xpack.security.http.ssl.enabled=false" \
  -e "xpack.license.self_generated.type=trial" \
  docker.elastic.co/elasticsearch/elasticsearch:8.15.0
```

and the link to go to the quick start guide to start with kibana and connect with a lang client of their choice.

Its not quite as nice as qdrant, but these params are important.

maxjakob commented on docs/reference/quickstart/run-elasticsearch-locally.asciidoc:
```diff
@@ -0,0 +1,177 @@
+  -e ELASTICSEARCH_HOSTS=http://elasticsearch:9200 \
+  -e ELASTICSEARCH_USERNAME=kibana_system \
+  -e ELASTICSEARCH_PASSWORD=$KIBANA_PASSWORD \
+  -e "xpack.security.enabled=false" \
```
Just to double-check and for me to learn: is this correct or should this be `xpack.security.http.ssl.enabled=false` to keep Kibana password as required?

miguelgrinberg replied:
The names of these options aren't consistent between Kibana and ES. For Kibana, this option disables TLS, but Kibana will still authenticate with its own password, and also ask the user to log in before providing access.
